### DISCUSSION TITLE
Sleep tracking

### DISCUSSION THREAD

**Discussion by 88572** [InfiniTimeOrg/InfiniTime#1797] [General]
Hello, I recently got a sealed PineTime. On the box, it says Sleep Monitor and Calories Burnt. When I turned it on, I couldn't find either feature. Was it removed?
Thanks!

**Comment by Elements6007** [comment]
Unfortunately the infinitime does not have those features yet. Although I did find the labels for sleep monitor and calorie counter on my box you are talking about.

**Comment by thiswillbeyourgithub** [comment]
I made a sleeptracker for waspos btw https://github.com/thiswillbeyourgithub/sleep_tracker_pinetime_wasp-os

**Reply by 88572** [reply]
Can we port this?

**Reply by thiswillbeyourgithub** [reply]
For anyone else wondering about sleep tracking on the pinetime: through the power of offering a bounty and luck of finding the awesome and very competent @cyberneel a port is in the way. It is currently a fork but the PR is ongoing at https://github.com/InfiniTimeOrg/InfiniTime/pull/2174

Let that be an inspiration for anyone to help financially incentivize opensource contributors!

**Comment by lxrst** [comment]
@thiswillbeyourgithub Are you ok with me porting your code to InfiniTime?

**Reply by thiswillbeyourgithub** [reply]
Please do! I still use it daily and have been for about 2 and a half years. It's one of the most useful thing I ever did :D. It's probably bug free. Though I stopped logging and use only the alarm now so there may be issues i'm not aware of.
I do ask that you credit me though :)
If you can port it in a way that still exposes the same parameters that I set at the top of the script, that would be nice.
If you find a bug or have any question don't hesitate. Maybe open an issue in my repo to make sure I see it.
You can update the logo if you want, if it could be reminiscent of the original that would be nice too :)

I have a decent chance of using your port because infinitime has so much more battery life :)

Good'day!

**Reply by thiswillbeyourgithub** [reply]
Also : there's [a branch](https://github.com/thiswillbeyourgithub/sleep_tracker_pinetime_wasp-os/tree/beared_dude_enhancement) you should know about. Basically one user who knows his stuff more than me encouraged me to rewrite some parts to use less memory and guided me a bit. IIRC it was very close to completion but I ran into weird issues I didn't have the time to figure out and fix, related to menu navigation (like scrolling back and forth didn't work as I wanted to) and probably other stuff as well. I don't think I changed the main branch at all for a long long time so that branch should be almost the same, except more memory intensive.

Maybe that branch would be a better inspiration for your port? I'll leave that up to you!

Have a nice day and do keep me updated :)!

**Reply by tituscmd** [reply]
> @thiswillbeyourgithub Are you ok with me porting your code to InfiniTime?

I'm also very interested in a sleep tracker for InfiniTime! Could you keep us up-to-date on how the porting is going?

**Reply by thiswillbeyourgithub** [reply]
(Also if the porting happens somewhere accessible I'll gladly add a link from my own repo to yours)

**Reply by thiswillbeyourgithub** [reply]
Note: I just pushed a fix to sleeptk: one of my pinetime seems to have a broken accelerometer that raises an exception when the accelerometer is used and I was wrongly resetting the accelerometer even if body trackign was disabled. When doing your port, please take into consideration broken hardware, ie don't do any call to hardware if it's not needed! With this fix all my watches can be used fine.

Thanks :)

**Reply by tituscmd** [reply]
> @thiswillbeyourgithub Are you ok with me porting your code to InfiniTime?

How's this coming along?

**Reply by tituscmd** [reply]
> @thiswillbeyourgithub Are you ok with me porting your code to InfiniTime?

It's now been about 5 months, has any porting been done? Not trying to sound demanding, I'm just really interested in a sleep tracker native on InfiniTime 😄

**Reply by thiswillbeyourgithub** [reply]
For anyone else wondering about sleep tracking on the pinetime: through the power of offering a bounty and luck of finding the awesome and very competent @cyberneel a port is in the way. It is currently a fork but the PR is ongoing at https://github.com/InfiniTimeOrg/InfiniTime/pull/2174

Let that be an inspiration for anyone to help financially incentivize opensource contributors!

**Comment by thiswillbeyourgithub** [comment]
### edit to update:
For anyone else wondering about sleep tracking on the pinetime: through the power of offering a bounty and luck of finding the awesome and very competent @cyberneel a port is in the way. It is currently a fork but the PR is ongoing at https://github.com/InfiniTimeOrg/InfiniTime/pull/2174

Let that be an inspiration for anyone to help financially incentivize opensource contributors!


---

Hi,

I'm the dev behind [SleepTk](https://github.com/thiswillbeyourgithub/SleepTk_pinetime_sleep_tracker) and have been using it over 99% of my nights **for a few years now**. It's one of my most useful project to my health. I'm actually not using it as a sleep tracker though, but using it to:
- gently raise to an awakened state via customizable gradually increasing vibrations
- wakeup feeling fresh thanks to keeping track of sleep cycle
- knowing more about my insomnia by knowing mid sleep how far into a cycle I'm into (+ a red display to avoid blue light) 
- wakeup without waking up my significant other
- wake up thanks to snooze hell (it's easy to snooze for 3min but hard to stop, making sure you won't get back to sleep)
- Sometimes use "natural wake" mode to vibrate just a bit every few seconds. This wakes me up only if I had enough rest, otherwise I can sleep through it

Overall I'm extremely satisfied with SleepTk and it dramatically improved my sleep hygiene but the only thing I'm missing is infinitime's battery life. Hence, I'm looking for anyone willing to port the code / make a similarly featured app to the infinitime. An added benefit is to increase the usability of those sleep features (I could then offer a pinetime with infinitime and sleeptk to people, as wasp-os is kinda unsuited for non dev and the low battery life is a real issue for people). **My friend and I made both a $50 bounty for anyone willing to do that. So at least $100 can go to you or the project of your choice if you decide to do it :)! Do reach out by creating an issue at [the sleeptk repo](https://github.com/thiswillbeyourgithub/SleepTk_pinetime_sleep_tracker) to talk details**.

# TL;DR: $100 bounty to a port of sleeptk's wakeup features to infinitime.

Extra pinging to @tituscmd  @lxrst  @everypizza1 as you were especially wondering about porting SleepTk to the infinitime.


edit: The centralized issue for the people interested in porting is [here](https://github.com/thiswillbeyourgithub/SleepTk_pinetime_sleep_tracker/issues/13)!

**Reply by Koloss0** [reply]
If this gets off the ground (fork or not) and implemented, I will add another $56 CAD (\~$40 USD). But there must be heart tracking, cycle calculating (based on alarm time and current time), and the gradual wake feature. The midnight stats screen will earn another $20 CAD (\~$14 USD). I think a separate fork of InfiniTime with a focus on sleep would be ideal so it could be extended by the community and new features can be implemented. Also yes, I'm willing to drop $76 CAD over this, I want open source sleep tracking without the battery cost of WaspOS.

**Reply by thiswillbeyourgithub** [reply]
Obvious conflict of interest but I've used SleepTk for years so I can totally vouch for it being worth every penny. That's why I figured I'd make a bounty, it's a great health investment given the number of night per day!

**Reply by thiswillbeyourgithub** [reply]
For anyone else wondering about sleep tracking on the pinetime: through the power of offering a bounty and luck of finding the awesome and very competent @cyberneel a port is in the way. It is currently a fork but the PR is ongoing at https://github.com/InfiniTimeOrg/InfiniTime/pull/2174

Let that be an inspiration for anyone to help financially incentivize opensource contributors!
